File: 4-Behavioral Patterns/5-Memento Pattern/Text Editor Example/Java/runner/TextEditorApp.java
package runner;

import text_editor.Caretaker;
import text_editor.Memento;
import text_editor.TextEditor;

import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JPanel;

/**
 * TextEditorApp class that works as "Caretaker" that actually uses Memento
 * Pattern.
 * This class is responsible for keeping a collection of "Memento", but never
 * examines or operates on the contents of a "Memento".
 *
 * @author Ziang Lu
 */
public class TextEditorApp extends JFrame {

    /**
     * Text editor of this application.
     */
    private TextEditor textEditor = new TextEditor(5, 10);
    /**
     * Caretaker that stores all the mementos.
     */
    private Caretaker caretaker = new Caretaker();
    /**
     * Pointer to the last saved state.
     */
    private int prevStatePtr = -1;

    /**
     * Default constructor.
     */
    public TextEditorApp() {
        // Set the basic information for this frame
        setSize(400, 200);
        setDefaultCloseOperation(EXIT_ON_CLOSE);
        setTitle("Memento Pattern demo");

        // Save for the initial empty text
        save();

        JPanel panel = new JPanel();
        panel.add(new JLabel("Text:"));
        panel.add(textEditor.getTextField());

        JButton saveButton = new JButton("Save");
        saveButton.addActionListener((e) -> { // Java lambda expression in place of anonymous class
            if (e.getSource() == saveButton) {
                save();
            }
        });
        panel.add(saveButton);

        JButton undoButton = new JButton("Undo");
        undoButton.addActionListener((e) -> {
            if (e.getSource() == undoButton) {
                undo();
            }
        });
        panel.add(undoButton);

        JButton redoButton = new JButton("Redo");
        redoButton.addActionListener((e) -> {
            if (e.getSource() == redoButton) {
                redo();
            }
        });
        panel.add(redoButton);

        add(panel);

        // Display this frame
        setVisible(true);
    }

    /**
     * Saves the current state of the text editor.
     */
    private void save() {
        caretaker.addMemento(textEditor.createMemento());
        prevStatePtr = caretaker.getLastMementoPtr();
    }

    /**
     * Undo the most recent change in the text editor.
     */
    private void undo() {
        if (!textEditor.getText().equals(caretaker.getMemento(prevStatePtr).getText())) {
            Memento lastSavedState = caretaker.getMemento(prevStatePtr);
            textEditor.restore(lastSavedState);
        } else if (prevStatePtr <= 0) {
            System.out.println("No more undo operation available!");
        } else {
            --prevStatePtr;
            Memento prevState = caretaker.getMemento(prevStatePtr);
            textEditor.restore(prevState);
        }
    }

    /**
     * Redoes the most recent change in the text editor.
     */
    private void redo() {
        if (prevStatePtr >= caretaker.getLastMementoPtr()) {
            System.out.println("No more redo operation available!");
            return;
        }
        ++prevStatePtr;
        Memento nextState = caretaker.getMemento(prevStatePtr);
        textEditor.restore(nextState);
    }

    /**
     * Main driver.
     * @param args arguments from command line
     */
    public static void main(String[] args) {
        new TextEditorApp();
    }

}

File: 4-Behavioral Patterns/7-Memento Pattern/Text Editor Example/Python/text_editor_app.py
#!usr/bin/env python3
# -*- coding: utf-*-

"""
Text editor application module.
"""

__author__ = 'Ziang Lu'

from text_editor import Memento, TextEditor


class TextEditorApp:
    """
    TextEditorApp class that works as "Caretaker" that actually uses Memento
    Pattern.
    This class is responsible for keeping a collection of "Memento", but never
    examines or operates on the contents of a "Memento".
    """
    __slots__ = ['_text_editor', '_memento_list', '_prev_state_ptr']

    def __init__(self):
        """
        Default constructor.
        """
        self._text_editor = TextEditor()
        self._memento_list = []
        self._prev_state_ptr = -1

    def write(self, text: str) -> None:
        """
        Writes the given text to the text editor.
        :param text: str
        :return: None
        """
        self._text_editor.text = text
        print(f"After writing '{text}': {self._text_editor}")

    def save(self) -> None:
        """
        Saves the current state of the text editor.
        :return: None
        """
        self._add_memento(self._text_editor.create_memento())
        self._prev_state_ptr = len(self._memento_list) - 1

    def _add_memento(self, memo: Memento) -> None:
        """
        Private helper function to store the given memento.
        :param memo: Memento
        :return: None
        """
        self._memento_list.append(memo)

    def undo(self) -> None:
        """
        Undo the most recent change in the text editor.
        :return: None
        """
        print('Undo operation...')
        if self._text_editor.text !=\
                self._memento_list[self._prev_state_ptr].text:
            last_saved_state = self._memento_list[self._prev_state_ptr]
            self._text_editor.restore(last_saved_state)
            print(self._text_editor)
        elif self._prev_state_ptr <= 0:
            print('No more undo operation available!')
        else:
            self._prev_state_ptr -= 1
            prev_state = self._get_memento(self._prev_state_ptr)
            self._text_editor.restore(prev_state)
            print(self._text_editor)

    def _get_memento(self, idx: int) -> Memento:
        """
        Private helper function to get the memento at the given index.
        :param idx: int
        :return: Memento
        """
        return self._memento_list[idx]

    def redo(self) -> None:
        """
        Redoes the most recent change in the text editor.
        :return: None
        """
        print('Redo operation...')
        if self._prev_state_ptr >= len(self._memento_list) - 1:
            print('No more redo operation available!')
            return
        self._prev_state_ptr += 1
        next_state = self._get_memento(self._prev_state_ptr)
        self._text_editor.restore(next_state)
        print(self._text_editor)
